packages/next/src/client/components/segment-cache/navigation.ts

```diff
@@ -0,0 +1,258 @@
+  // use the same server response to write the actual data into the CacheNode
+  // tree. So it's the same flow as the "happy path" (prefetch, then
+  // navigation), except we use a single server response for both stages.
+
+  const promiseForDynamicServerResponse = fetchServerResponse(url, {
+    flightRouterState: currentFlightRouterState,
+    nextUrl,
+  })
+  const { flightData } = await promiseForDynamicServerResponse
+
+  // TODO: Detect if the only thing that changed was the hash, like we do in
+  // in navigateReducer
+
+  if (typeof flightData === 'string') {
+    // This is an MPA navigation.
+    const newUrl = flightData
+    return {
+      tag: NavigationResultTag.MPA,
+      data: newUrl,
+    }
+  }
+
+  // Since the response format of dynamic requests and prefetches is slightly
+  // different, we'll need to massage the data a bit. Create FlightRouterState
+  // tree that simulates what we'd receive as the result of a prefetch.
+  const prefetchFlightRouterState = simulatePrefetchTreeUsingDynamicTreePatch(
+    currentFlightRouterState,
+    flightData
+  )
+
+  // In our simulated prefetch payload, we pretend that there's no seed data
+  // nor a prefetch head.
+  const prefetchSeedData = null
+  const prefetchHead = null
+  const task = updateCacheNodeOnNavigation(
+    currentCacheNode,
+    currentFlightRouterState,
+    prefetchFlightRouterState,
+    prefetchSeedData,
+    prefetchHead
+  )
+
+  // Now we proceed exactly as we would for normal navigation.
+  if (task !== null) {
+    const newCacheNode = task.node
+    if (newCacheNode !== null) {
+      listenForDynamicRequest(task, promiseForDynamicServerResponse)
+    }
+    return {
+      tag: NavigationResultTag.Success,
+      data: {
+        flightRouterState: task.route,
+        cacheNode: newCacheNode !== null ? newCacheNode : currentCacheNode,
+      },
+    }
+  }
+
+  // The server sent back an empty tree patch. There's nothing to update.
+  return noOpNavigationResult
+}
+
+function simulatePrefetchTreeUsingDynamicTreePatch(
+  currentTree: FlightRouterState,
+  flightData: Array<NormalizedFlightData>
+): FlightRouterState {
+  // Takes the current FlightRouterState and applies the router state patch
+  // received from the server, to create a full FlightRouterState tree that we
+  // can pretend was returned by a prefetch.
+  //
+  // (It sounds similar to what applyRouterStatePatch does, but it doesn't need
+  // to handle stuff like interception routes or diffing since that will be
+  // handled later.)
+  let baseTree = currentTree
+  for (const { segmentPath, tree: treePatch } of flightData) {
+    // If the server sends us multiple tree patches, we only need to clone the
+    // base tree when applying the first patch. After the first patch, we can
+    // apply the remaining patches in place without copying.
+    const canMutateInPlace = baseTree === currentTree
+    simulatePrefetchTreeUsingDynamicTreePatchImpl(
```

Comment:
I realize this is still a draft, but I wanted to jot this down before I forget. Based on the comment here, shouldn't the code be like this instead?

```suggestion
  let baseTree = currentTree
  for (const { segmentPath, tree: treePatch } of flightData) {
    // If the server sends us multiple tree patches, we only need to clone the
    // base tree when applying the first patch. After the first patch, we can
    // apply the remaining patches in place without copying.
    const canMutateInPlace = baseTree !== currentTree
    baseTree = simulatePrefetchTreeUsingDynamicTreePatchImpl(
```

Reply:
Yeah my bad